Feature #28 구매 기능 구현

sunggwon opened the pull request:
사용자가 구매 진행하여 주문 관련 테이블에 데이터를 삽입하는 기능

sunggwon committed:
구매 기능 기본 구조 설정

sunggwon committed:
구매 기본 기능 구현 완료

sunggwon commented:
@f-lab-robert @KJW921222 
주원님과 리뷰 후 다음과 같은 부분을 개선사항을 확인하였습니다.
  - Response에 대한 응답 처리 추가
  - DTO 유효성 검사
  - 클래스 네이밍 변경
      - PurchaseDTO -> OrdersItemDTO로 변경
  - POJO 생성자 -> Builder 패턴으로 변경 (Insert에 사용할 경우)

sunggwon commented on src/main/java/com/flabedu/small/small/dto/PurchaseDTO.java:
```diff
@@ -0,0 +1,13 @@
+
+@Getter
+@Setter
+public class PurchaseDTO {
```
클래스 이름 변경 PurchaseDTO -> OrdersDTO

f-lab-robert commented on src/main/java/com/flabedu/small/small/TestController.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class TestController {
```
PR 을 올리기 전에 Test 관련 코드는 다 제거해주세요

sunggwon replied:
@f-lab-robert  프로젝트에서 테스트한 내용은 삭제하였습니다. 그런데 junit으로 단위테스트를 진행한 부분도 pr에서 제외 시키는게 좋을까요? 작성해둔 테스트를 머지 한 후 재 테스트 할 수 있지 않나요?

f-lab-robert replied:
>  junit으로 단위테스트를 진행한 부분도 pr에서 제외 시키는게 좋을까요? 작성해둔 테스트를 머지 한 후 재 테스트 할 수 있지 않나요?
테스트 코드와 개인의 테스트를 위해 임의로 만든 코드를 제거하는 건 다른 의미 입니다.
GitHub repo 는 차후에 10명이상도 함께 작업할 수 있는 함께 협력해야 하는 코드만이 존재해야 합니다. TestController 는 성권님의 개인 테스트를 위해 작성한게 아닌가요 ? 이 부분을 제거해야 된다는 겁니다. 

다시 한번 말씀드리지만, 테스트 코드는 다른 의미입니다. 테스트 코드는 Junit, Mockito etc 을 의미합니다.

sunggwon replied:
코드상에 있는 테스트 관련 코드 제거하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/ItemController.java:
```diff
@@ -0,0 +1,28 @@
+@RestController
+public class ItemController {
+
+
```
불필요한 공백이 많습니다. 제거해주세요. 나아가 코드 정렬도 확인해봐주세요

sunggwon replied:
불필요한 공백을 제거하고 공백에 대한 일관성을 유지하도록 수정하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/ItemController.java:
```diff
@@ -0,0 +1,28 @@
+
+    public ItemController(ItemService service) {
+        this.service = service;
+    }
```
lombok 을 찾아보신 후에 적절한 기능 찾아서 활용해주세요. 위와 같은 코드는 좋지 않습니다.

sunggwon replied:
lombok과 생성자 주입을 사용하도록 수정하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/ItemController.java:
```diff
@@ -0,0 +1,28 @@
+    @PostMapping("/items/product")
+    public void purchaseItem(@RequestBody PurchaseDTO purchaseItem, HttpSession session){
+        service.purchaseItem(purchaseItem.getOrders(), session.getAttribute("UserID").toString());
+    }
```
저희가 HTTP 응답 값을 배웠던 부분 기억하시나요 ? 
성공하면 201, 실패하면 그에 적절한 Error 를 제공해줘야겠죠 ? 
우선 ResponseEntity 에 대해 학습해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/ItemController.java:
```diff
@@ -0,0 +1,28 @@
+
+    @ResponseBody
+    @PostMapping("/items/product")
+    public void purchaseItem(@RequestBody PurchaseDTO purchaseItem, HttpSession session){
```
1. DTO valid 체크가 전혀 안되어있네요. API 를 호출하는 쪽에서 API 규격을 따르지 않았을 경우에는 Error 가 어디서 나는지 찾기가 굉장히 힘들 것 같습니다. 나아가 검증되지 않은 input 으로 sql 을 날린다는건 어마어마한 보안상의 문제가 될 것 같고요. 

2. 우선은 Session 부분을 제거해주시고 사용자의 정보를 DB 에서 불러와서 insert 할 때 활용하시겠어요 ? PurchaseDTO 에 memberId 가 있다면 가장 좋을 것 같고요. 

3. 회원가입과 그에 따른 로직은 현 단계에서는 고려하지 않습니다. 차후에 refactoring 의 형태로 진행하는게 가장 좋을 것 같습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/LoginBlockFilter.java:
```diff
@@ -0,0 +1,33 @@
+import java.io.IOException;
+
+//@Component
+public class LoginBlockFilter implements Filter {
```
session 과 그에 따른 기능들은 현단계에서 고려하지 않습니다. 본 issue 와 맞지 않는 task 이고요. 

때문에 이 관련 부분들 모두 제거해주세요.

sunggwon replied:
제거하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -0,0 +1,84 @@
+        this.itemDetailMapper = itemDetailMapper;
+        this.ordersMapper = ordersMapper;
+        this.ordersItemMapper = ordersItemMapper;
+    }
```
lombok 활용해주세요. 굉장히 불완전한 코드 형태 입니다.

sunggwon replied:
lombok을 사용하도록 변경하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/repository/OrdersMapper.java:
```diff
@@ -0,0 +1,16 @@
+import org.apache.ibatis.annotations.Options;
+import org.apache.ibatis.annotations.Select;
+
+@Mapper
```
mapper, repository 어노테이션의 역할이 무엇인지 학습해주세요

f-lab-robert commented on src/main/java/com/flabedu/small/small/repository/OrdersMapper.java:
```diff
@@ -0,0 +1,16 @@
+public interface OrdersMapper {
+    @Insert("INSERT INTO orders " +
+            "(member_id, total_price,  status,orders_date, modified_date) " +
+            "values(#{memberId}, #{totalPrice}, #{status}, #{ordersDate}, #{modifiedDate})")
```
xml 로 빼서 진행해보시겠어요 ? Java 코드에서는 Java 코드만을 다루는게 가장 좋은 책임/역할 분리의 시작입니다.

sunggwon replied:
쿼리 관련 코드는 xml로 분리하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/repository/OrdersMapper.java:
```diff
@@ -0,0 +1,16 @@
+            "(member_id, total_price,  status,orders_date, modified_date) " +
+            "values(#{memberId}, #{totalPrice}, #{status}, #{ordersDate}, #{modifiedDate})")
+    @Options(useGeneratedKeys = true, keyProperty = "orderId")
+    void insertOrders(Orders orders);
```
option 은 어떤 목적으로 사용하신걸까요

sunggwon replied:
@f-lab-robert 
Orders와 Orders_item이 같이 삽입 되어야하는데 orders_item이 order의 auto_increment 속성의 ID를 외래키로 가지고 있어서 Orders를 삽입 후 ID 를 읽어오기 위해서 사용하였습니다.

![image](https://user-images.githubusercontent.com/17060341/169245311-345a0a72-7ccc-4e3a-820d-6d9e63c8bae1.png)

f-lab-robert commented on src/main/java/com/flabedu/small/small/FilterConfiguration.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class FilterConfiguration {
```
1. session 으로 인해 만드신 거라면 제거해주세요.
2. config 를 차후에 다루실 일이 있다면 config 폴더를 만드는게 적합해보입니다.

sunggwon replied:
제거하였고 configuration 패키지를 따로 생성하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/TestController.java:
```diff
@@ -0,0 +1,18 @@
+public class TestController {
+
+    @Autowired
+    MemberMapper memberMapper;
```
Autowired 는 안좋은 주입 형태 입니다. 저희가 예전에 스프링 학습했던거를 기억해주세요

sunggwon replied:
생성자 주입으로 변경하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -0,0 +1,84 @@
+        this.ordersItemMapper = ordersItemMapper;
+    }
+
+    @Transactional
```
Transactional 의 전파레벨, readonly 등에 대해 학습해주세요. 그리고 정리해서 리뷰 요청 해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -0,0 +1,84 @@
+        Member user = memberMapper.getMember(userID);
+        List<OrdersItem> ordersItems = createOrderItems(user, itemInfo);
+        ordersItems.forEach(
+                v-> System.out.println(v.getItemId())
```
lombok 에 대해 학습해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -0,0 +1,84 @@
+
+            if(item == null) throw new IllegalArgumentException("존재하지 않은 Item");
+            if(detail == null) throw new IllegalArgumentException("존재하지 않은 아이템 사이즈");
+            if(detail.getStock() <= 0) throw new IllegalStateException("재고 없음");
```
이렇게 Error 를 사용자에게 주는 경우는 없습니다. CustomException 을 만드는 방식에 대해 찾아봐주세요.

f-lab-robert commented:
@f-lab-robert @KJW921222
주원님과 리뷰 후 다음과 같은 부분을 개선사항을 확인하였습니다.

Response에 대한 응답 처리 추가
DTO 유효성 검사
클래스 네이밍 변경
PurchaseDTO -> OrdersItemDTO로 변경
POJO 생성자 -> Builder 패턴으로 변경 (Insert에 사용할 경우)

--> 현 단계부터는 금주 동안 두분이 상의하지 말아주세요. 아무 리뷰도 받지 않은 상태에서 잘못된 논의들이 이루어지고 있는 것 같습니다. 재차 말씀드렸듯이 두 분이 먼저 각각 API 를 개발하고 "멘토" 에거 리뷰를 요청해주세요. 현재는 개발을 제대로 시작하기 전입니다. 때문에 "스스로 개발" -> "멘토에게 리뷰 요청" -> "다른 멘티의 코드 확인 및 상의"  순으로 이루어져야 합니다. "스스로 개발" <-> "멘티와 상의" 이 사이클로 이루어지면 아무런 리뷰를 받지 못한 상태에서 코드의 양과 불필요한 논의만 많아지는 형태입니다.

f-lab-robert commented:
[ 폴더 구조 ]
- web layer 는 기본적으로 dto 를 포함합니다. 그게 아니라면 controller 라는 폴더명이 더 어울리는 형태인 것 같습니다.
- dto 에는 request (요청) 과 response (응답) 으로 구분이 되어야할 것 같습니다.
- repository 밑에 mapper 가 있는건 이질감이 느껴집니다. repository 와 mapper 어노테이션 역할에 대해 찾아보고 리뷰요청해주세요.
- dao 의 역할이 무엇인지 다시한번 생각해주세요. 
- config 에 대해서는 별도의 config 폴더를 만들어주셔도 좋을 것 같습니다. 

[ 기타 ]
- test code 를 작성하지 않은 상태로 이렇게 많은 코드를 만드는건 올지 않습니다. 스스로 로직 하나하나 작성할 때마다 test code 를 작성해주세요. 
- 현재 사용하고 계신 DTO 는 DTO 라기 보다 Model 에 가까운 형태로 사용하고 있는 것 같습니다. DTO 명을 requestDTO 와 같은 형태로 바꿔주시고. 다른 DTO 1개가 다른 DTO 에 의존을 갖지 않게 해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/data/Item.java:
```diff
@@ -0,0 +1,21 @@
+    String registerUserID;
+    Date registerDate;
+    String modifiedUserID;
+    Date modifiedDate;
```
LocalDateTime 에 대해 찾아보세요. 저희가 Java 의 신에서 학습했던 내용인것 같습니다.

sunggwon replied:
Date 대신 LocalDateTime을 적용하였습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/data/Orders.java:
```diff
@@ -0,0 +1,28 @@
+    long orderId;
+    long memberId;
+    long totalPrice;
+    String status = "S";
```
Enum 에 대해 찾아봐주세요

sunggwon committed:
Mapper 코드 xml 로 분리

sunggwon committed:
- Mapper 테스트 작성 및 xml과 분리
- DTO 의존성 제거
- Enum 적용
- Custom Exception 정의 후 Controller에서 예외 처리 후 HTTP 상태 메시지 전달

sunggwon committed:
- @Data lombok 어노테이션 적용

f-lab-robert commented on src/main/java/com/flabedu/small/small/SMallApplication.java:
```diff
@@ -6,6 +6,7 @@
 @SpringBootApplication
 public class SMallApplication {
 
+
```
불필요한 공백입니다. 신경써주세요.

f-lab-robert commented:
우선 아래 피드백을 참고해주시고요... 말씀드릴 내용이 너무나 많기에 자세한 부분은 멘토링 시간에 이야기 하도록 할게요.

[ 피드백 ]

- 마인드셋
1. 혼자 개발하는게 아닙니다. TODO 라고 적은 내용은 다 학습을 하기 전까지는 PR 을 올리시면 안됩니다. 개발자 스스로가 잘 알지도 못하고 사용하는 코드를 몇십만명이 사용하는 유저에게 노출된다고 생각해보세요.

- 폴더 구조 
2. 저희 계층에서 data 라는 폴더를 사용하기로 한적 이 없습니다. 왜 사용하신걸까요.
3. 왜 서비스를 인터페이스와 구현체로 구분해서 사용하셨을까요 ? 굉장히 오래전에 관습적으로 사용되는 코딩 패턴인데 어떤 근거와 어떤 코드를 참고해서 이렇게 사용하신걸까요 ?

- 코딩 
4. 조금 더 하나에 대해 깊게 생각해보시는 습관이 필요할 것 같습니다. 예컨대 Spring 예외처리에 대해 구글링해보셨나요 ? 그렇다면 RestControllerAdivce 를 발견하실 수 밖에 없었으리라 생각이 듭니다. 이 부분을 학습해보시겠어요 ? 재차 말씀드리지만 대충 코딩하는게 아니라 하나하나 천천히 해나가야 됩니다. 
5. lombok 도 마찬가지입니다. AllArgsConstructor, Data 를 마구잡이로 사용한 느낌이 강합니다. 생성자 주입이 좋은 이유에 대해 예전에 학습하셨던거 기억하실까요 ? 그렇다면 그 부분을 중점으로 어떻게 lombok 과 엮어서 사용하는게 best practice 인지 조금만 구글링 혹은 다른 프로젝트를 참고해도 알 수 있습니다. 옆을 보지 않고 앞만 보고 달리고 있는 느낌이 강합니다. 재차 말씀드리지만 하나하나 꼼꼼하게 해주세요. lombok 에 대해서도 전혀 학습이 되어있지 않는 것 같습니다.
6. mapper/repository 의 어노테이션 의미에 대해서도 찾아보시지 않은것 같습니다. 나아가 repository 하위에 mapper 를 둔 이유에 대해 잘 모르겠습니다. 설명이 필요할 것 같습니다.
7. dto validation 도 마찬가지입니다. 이 API 를 사용하는 사용자 입장에서 생각해봐주세요. 어떠한 형태의 에러가 생기던 똑같은 format 으로 사용자에게 전달되는지 확인해보셨나요 ? 에러의 형태가 A 일때랑 B 상황일 떄 모두 다르면 안드로이드/웹 개발자들은 어떻게 포맷팅을 해서 에러를 관리할 수 있을까요 ?
8. 계층형 폴더 구조의 정의를 명확하게 생각해봐주세요. service error 가 왜 controller 에서 처리가 되고 있는걸까요 ? 왜 try catch 문이 있는형태죠 ?
9. 불필요한 공백, 코드 정렬이 너무나 많습니다. 다 같이 개발하는 코드 입니다. 20명이 모두 본인이 원하는 곳에 1,2 줄씩 맘대로 띄우거나 코드정렬을 마음대로 하는건 올바르지 못합니다.
10. 다른 프로젝트의 테스트 코드를 참고해봐주세요. 테스트 코드에 대한 머릿속 그림이 아직 많이 부족해보입니다.

sunggwon committed:
Exception 처리 관런 RestControllerAdvice 핸들러를 사용하여 전역적으로 처리하도록 변경

JinXuanT commented:
> 우선 아래 피드백을 참고해주시고요... 말씀드릴 내용이 너무나 많기에 자세한 부분은 멘토링 시간에 이야기 하도록 할게요.
> 
> [ 피드백 ]
> 
>     * 마인드셋
> 
> 
>     1. 혼자 개발하는게 아닙니다. TODO 라고 적은 내용은 다 학습을 하기 전까지는 PR 을 올리시면 안됩니다. 개발자 스스로가 잘 알지도 못하고 사용하는 코드를 몇십만명이 사용하는 유저에게 노출된다고 생각해보세요.
> 
> 
>     * 폴더 구조
> 
> 
>     2. 저희 계층에서 data 라는 폴더를 사용하기로 한적 이 없습니다. 왜 사용하신걸까요.
> 
>     3. 왜 서비스를 인터페이스와 구현체로 구분해서 사용하셨을까요 ? 굉장히 오래전에 관습적으로 사용되는 코딩 패턴인데 어떤 근거와 어떤 코드를 참고해서 이렇게 사용하신걸까요 ?
> 
> 
>     * 코딩
> 
> 
>     4. 조금 더 하나에 대해 깊게 생각해보시는 습관이 필요할 것 같습니다. 예컨대 Spring 예외처리에 대해 구글링해보셨나요 ? 그렇다면 RestControllerAdivce 를 발견하실 수 밖에 없었으리라 생각이 듭니다. 이 부분을 학습해보시겠어요 ? 재차 말씀드리지만 대충 코딩하는게 아니라 하나하나 천천히 해나가야 됩니다.
> 
>     5. lombok 도 마찬가지입니다. AllArgsConstructor, Data 를 마구잡이로 사용한 느낌이 강합니다. 생성자 주입이 좋은 이유에 대해 예전에 학습하셨던거 기억하실까요 ? 그렇다면 그 부분을 중점으로 어떻게 lombok 과 엮어서 사용하는게 best practice 인지 조금만 구글링 혹은 다른 프로젝트를 참고해도 알 수 있습니다. 옆을 보지 않고 앞만 보고 달리고 있는 느낌이 강합니다. 재차 말씀드리지만 하나하나 꼼꼼하게 해주세요. lombok 에 대해서도 전혀 학습이 되어있지 않는 것 같습니다.
> 
>     6. mapper/repository 의 어노테이션 의미에 대해서도 찾아보시지 않은것 같습니다. 나아가 repository 하위에 mapper 를 둔 이유에 대해 잘 모르겠습니다. 설명이 필요할 것 같습니다.
> 
>     7. dto validation 도 마찬가지입니다. 이 API 를 사용하는 사용자 입장에서 생각해봐주세요. 어떠한 형태의 에러가 생기던 똑같은 format 으로 사용자에게 전달되는지 확인해보셨나요 ? 에러의 형태가 A 일때랑 B 상황일 떄 모두 다르면 안드로이드/웹 개발자들은 어떻게 포맷팅을 해서 에러를 관리할 수 있을까요 ?
> 
>     8. 계층형 폴더 구조의 정의를 명확하게 생각해봐주세요. service error 가 왜 controller 에서 처리가 되고 있는걸까요 ? 왜 try catch 문이 있는형태죠 ?
> 
>     9. 불필요한 공백, 코드 정렬이 너무나 많습니다. 다 같이 개발하는 코드 입니다. 20명이 모두 본인이 원하는 곳에 1,2 줄씩 맘대로 띄우거나 코드정렬을 마음대로 하는건 올바르지 못합니다.
> 
>     10. 다른 프로젝트의 테스트 코드를 참고해봐주세요. 테스트 코드에 대한 머릿속 그림이 아직 많이 부족해보입니다.

hey come down bro 😎

JinXuanT commented:
robert 형 빡쳤잖아 잘좀해라 얘들아

sunggwon committed:
컨트롤러 테스트 코드 작성
서비스 예외 테스트 코드 작성

f-lab-robert commented:
[2022.05.22]

1. 이전 피드백에 대한 내용들을 조금 더 신중하게 다뤄주세요. 

- domain 이라는 형태를 포기하고 계층형이라는 형태의 구조를 취하기로 했습니다. 때문에, domain 이라는 폴더명이 아닌 조금 더 계층 취지에 맞는 형태의 폴더 구조를 가져가기로 했던 것 같습니다. 이에 대한 논의의 연장선으로 model, dao 등에 대한 이야기가 오갔던 걸로 기억하는데요. 자의적인 판단이 아닌 기존에 내린 결정과 논의된 내용을 바탕으로 수정해주세요.

2. 다른 코드를 적극적으로 참고해주세요. ( 정말 많이 반복하고 있는 것 같습니다 )

- 가령, bean 주입 시 조사 결과 final ItemRepository itemRepository;  -> private final ItemRepository itemRepository; 의 형태가 더 많지 않았나요 ? 왜 그런걸까요 ?  f-lab [여러 코드들과](https://github.com/f-lab-edu/make-delivery/blob/75aff9c6fdcb049e9eccca9e979fa9f8c6cef57f/src/main/java/com/flab/makedel/controller/DeliveryController.java#L22) 여러 구글링을 통한 예제 코드를 통해 private final 을 통해 주입 받는 이점에 대해 확실히 정리를 하신걸까요 ? 

- 폴더명에 "언더바" 를 사용하는건 어떤 기준일까요 ? 다른 코드들에서는 어떤식으로 사용하고 있나요 ? 왜 그런걸까요 ? 

3. 비슷한 코드가 반복되는것 같으면 어떻게 하면 더 효율적으로 할 수 있을지 고민하고 찾아봐주세요. 

- 예컨대, customerorr 를 처리하려는 코드들은 내용은 똑같고 클래스명만 다릅니다. 비효율적으로 느껴지시지는 않았나요 ? 다른 프로젝트들의 예시들을 찾아보셨나요 ? [좋은 예시](https://github.com/f-lab-edu/kream/pull/76)

4. 저희가 사용하고 있는 기술에 초점을 맞춰주세요.

- 현재 구조상 repository 에서 사실상 mapper 를 한번 더 감싸서 service 에 넘겨주고 있습니다. 그렇다면, repository 가 없어도 별 문제가 없는 것 처럼 보이는데요. 저번에 service interface, service implement 를 만드셔서 사용하신 것과 동일해보입니다. 어떠한 이유가 있어서 이렇게 설계를 하신 걸까요 ?

5. 천천히 하나하나 학습하면서 진행해주세요. ( 똑같은 피드백이 너무나 반복되고 있습니다 )

- NotEmpty, NotNull 을 같이 사용해주셨는데. NotEmpty 가 null check 를 포함하고 있지는 않나요 ? NotBlank 에 대해서도 찾아보셨나요 ? 
- Enum 에 대해 더 학습해봐주세요. Enum 무분별하게 설계하고 이에 대한 정보를 가져오는 것 또한 기존에 있는 함수들로 활용할 수 있지 않나 찾아보는게 중요할 것 같습니다.
- 기타 등등 더 빠트린 내용이 없는지 찾아봐주세요.

6. 현 단계에서 추상화를 하려고 하지 말아주세요. 

- 가령, ErrorCodes 의 인터페이스가 있고 그 이하 상속 받는 클래스들이 몇 보이는데요. 현재 단계에서는 추상화를 하려고 시도하지 말아주세요. 가장 뼈대가 되는 기본 문제들에 집중을 하는게 중요한 것 같습니다.

7. 파일 명칭에 신경써주세요. 

- 주문에 대한 request 를 받는 dto 의 명칭이 ItemsProductDTO 일까요 ? 번역하면 "상품 상품 dto" 인데 어색하지 않나요 ? "OrderRequestDTO" 의 명칭이 조금 더 적합해보입니다. 무엇보다도 ItemsProductDTO 를 직역해보시겠어요 ? ( "상품 상품 DTO" )
- controller 명칭도 구매(주문)에 해당하는 기능이라면, Purchase (구매) // Order (주문) controller 가 더 적합하지 않을까요 ? 
- 그 외 등등

8. PR 전에 코드 점검을 몇번이고 더 해주세요.

```
    @NotNull
    @NonNull
    @NotEmpty(message = "orders must not be empty")
    List<OrderItem> orders = new LinkedList<>();
```

- @NotNull 을 2번씩이나 사용하고 있네요. 더불어 저번 PR 리뷰에서도 말씀드렸던 불필요한 띄어쓰기와 공백이 없는지도 봐주세요.

9. 기준을 명확하게 정하고 작업해주세요. 

- 어떤 TestCode 에서는 DisplayName 가 있고 어떤 TestCode 에서는 DisplayName 가 보이지 않네요. 

10. service 로직에 대한 test code 도 작성해주세요. 

- 위의 내용을 모두 수정하면 test code 에 대한 변경사항도 필요하리라 생각합니다. 그 부분들 먼저 수정해주세요. 
- service 로직에 대한 test code 도 작성해주세요. [좋은 예시](https://github.com/f-lab-edu/sns-itda/blob/develop/src/test/java/me/liiot/snsserver/service/AlarmServiceTest.java)

sunggwon committed:
- 커스텀 예외 통일
- 필드들 private 접근자로 지정
- controller dto 예외 처리
- 중복된 dto 예외 처리 구문 제거

sunggwon committed:
- 커스텀 예외 통일
- 필드들 private 접근자로 지정
- controller dto 예외 처리
- 중복된 dto 예외 처리 구문 제거

sunggwon committed:
- service 정상 동작 테스트 추가

sunggwon committed:
- 테스트 DisplayName 추가

sunggwon committed:
오타 제거

sunggwon committed:
Repository 어노테이션 지정

f-lab-robert commented on src/main/java/com/flabedu/small/small/exception/handler/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class ExceptionHandlerController extends ResponseEntityExceptionHandler {
```
1. ExceptionHandlerController 의 명칭을 CustomExceptionHandler 로 바꿔보시겠어요 ? controller 명칭이 어색해보입니다.
2. ResponseEntityExceptionHandler 를 상속받지 않아도 문제가 없어보이는데요. 


```
CustomException.java

@Getter
public class CustomException extends RuntimeException{

    CustomExceptionErrorList errorList;

    public CustomException(CustomExceptionErrorList errorList){
        this.errorList = errorList;
    }
}
```

```
CustomExceptionHandler.java  

@RestControllerAdvice
public class CustomExceptionHandler {
  .... 
}
```

3. 위와 같이 "에러" 와 "에러를 처리하는 핸들러" 를 분명하게 구분해보시는게 어떨까요

sunggwon committed:
CustomExceptionHandler 이름 변경 및 불필요한 상속 제거
